From 45bd762d8b188abb06ca0eeb0b3f5c44ed88ae7b Mon Sep 17 00:00:00 2001
From: Hien Huynh <hien.huynh.px@renesas.com>
Date: Wed, 8 May 2019 19:21:30 +0700
Subject: [PATCH 056/328] drm: rcar-du: Add support for swapping Odd/even data
 in dual-link mode

Based on Hardware Manual, bit 1st of LVDSTRIPE register support for
swapping odd/even data for LVDS port in dual-link mode.
This patch enables it for swapping requirement from DT side.

Signed-off-by: Hien Huynh <hien.huynh.px@renesas.com>
---
 drivers/gpu/drm/rcar-du/rcar_lvds.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/rcar-du/rcar_lvds.c b/drivers/gpu/drm/rcar-du/rcar_lvds.c
index 57d7084..a7894cc 100644
--- a/drivers/gpu/drm/rcar-du/rcar_lvds.c
+++ b/drivers/gpu/drm/rcar-du/rcar_lvds.c
@@ -81,6 +81,7 @@ struct rcar_lvds {
 	enum rcar_lvds_mode mode;
 	enum rcar_lvds_link_mode link_mode;
 	u32 id;
+	bool st_swap;	/*Odd/even Data swap mode in Dual-link LVDS */
 };
 
 #define bridge_to_rcar_lvds(bridge) \
@@ -384,13 +385,15 @@ static void rcar_lvds_dual_mode(struct rcar_lvds *lvds0,
 			LVDCTRCR_CTR2SEL_DISP | LVDCTRCR_CTR1SEL_VSYNC |
 			LVDCTRCR_CTR0SEL_HSYNC);
 	rcar_lvds_write(lvds0, LVDCHCR, lvdhcr);
-	rcar_lvds_write(lvds0, LVDSTRIPE, LVDSTRIPE_ST_ON);
+	rcar_lvds_write(lvds0, LVDSTRIPE, LVDSTRIPE_ST_ON
+				| (lvds0->st_swap ? LVDSTRIPE_ST_SWAP : 0));
 
 	rcar_lvds_write(lvds1, LVDCTRCR, LVDCTRCR_CTR3SEL_ZERO |
 			LVDCTRCR_CTR2SEL_DISP | LVDCTRCR_CTR1SEL_VSYNC |
 			LVDCTRCR_CTR0SEL_HSYNC);
 	rcar_lvds_write(lvds1, LVDCHCR, lvdhcr);
-	rcar_lvds_write(lvds1, LVDSTRIPE, LVDSTRIPE_ST_ON);
+	rcar_lvds_write(lvds1, LVDSTRIPE, LVDSTRIPE_ST_ON
+				| (lvds1->st_swap ? LVDSTRIPE_ST_SWAP : 0));
 
 	/* Set the LVDS mode and select the input. */
 	lvdcr0 = lvds0->mode << LVDCR0_LVMD_SHIFT;
@@ -840,6 +843,11 @@ static int rcar_lvds_parse_dt(struct rcar_lvds *lvds)
 		lvds->link_mode = RCAR_LVDS_SINGLE;
 	}
 
+	if ((lvds->info->quirks & RCAR_LVDS_QUIRK_DUAL_LINK) &&
+				lvds->link_mode == RCAR_LVDS_DUAL)
+		lvds->st_swap = of_property_read_bool(lvds->dev->of_node,
+								"swap");
+
 done:
 	of_node_put(local_output);
 	of_node_put(remote_input);
-- 
1.9.1

